### PR TITLE
feat: promote tool-derived diffs to first-class chat cards

### PR DESCRIPTION
**Pull request by CrysisDeu**
## Summary

Adopts the structured-diff display model (RFC: `docs/request-for-change/rfc-tool-derived-diff-cards.md`, included in this PR): kiro-cli's edit tools already deliver a structured diff on the ACP `tool_call` event, and the backend already renders it to a unified-diff string on `tool_input` — but the dashboard hid it inside the collapsed tool-call details panel while the prompt ordered the model to hand-write the same diff as a ```diff block, on every file change, no exceptions.

This PR makes the tool-derived diff the primary display and relaxes the prompt mandate to a runtime-conditional fallback.

### Changes

- **Frontend**: `ToolCallLine` renders an always-visible `DiffBlock` card below the pill when the tool kind is `edit` and its input is a unified diff. New `pages/chat/toolDiff.ts` owns detection: kind gate first (a shell command with diff-shaped input never promotes), 400-line cap (whole-file creates keep the collapsed rendering). `TurnBlock` excludes diff-card rows from the "Worked through N steps" collapse — same predicate family as workflow_run / spawn_run / MCP-app rows.
- **Backend**: `make_unified_diff`'s transport cap is raised from 6 000 chars to 64 KiB (sized so the full-card range is never cut; ~11x larger persisted `meta.input` / WS payloads for large edits, within the existing 1 MB tool-field ceiling), with line-boundary truncation and a machine-detectable annotation. `derive_edit_diff` turns bare-JSON edit args (create / insert / strReplace) into a unified diff server-side, so every `tool_input` consumer sees the same diff.
- **Backend**: `_tool_meta` persists the ACP tool `kind` so historical rows gate identically to live ones. Rows persisted before the field never promote (fail-safe).
- **Prompt**: the diff-block rule in `prompt.md`, `prompt-orchestrator.md`, and `_CRITICAL_RULES` becomes conditional — ```diff blocks are required only when (a) the change was made outside file-editing tools (sed, scripted bulk edits, git apply), or (b) the runtime is not the dashboard. Slack/Discord/Telegram renderers never see `tool_input` (verified: `messaging/driver.py` copies four fields onto OutputEvent), so message text remains their only file-change display and the mandate stays for them.

### Evidence

Isolated dev-gateway session showing an edit turn (fs_write, kind=edit): the card renders and stays visible with the turn collapsed. The shell negative case (kind=execute never promotes, even with diff-shaped input) is pinned by automated tests in toolDiff.test.tsx and TurnBlock.test.tsx.

| Light | Dark |
|---|---|
| light | dark |

### Tested

- New: `toolDiff.test.tsx` (13 cases: kind gating, shell false-positive rejection, line cap, historical-meta path, card render), `TurnBlock.test.tsx` diff-card visibility x3, backend `test_tool_meta_persists_kind`.
- Full gates: tsc, vitest 22747 passed (one failing file, `CliPanelCoverage`, passes standalone and pre-dates this change — full-run cross-file pollution), pytest 60119 passed (21 failures reproduce identically on pristine `origin/main` — host-environment class), flake8, mypy, docs-lint, brand gate all green. isort failure on `validation.py` also reproduces on pristine main.

### Not in scope

Extending messaging `OutputEvent` with `tool_input` so channel renderers could show diffs (would let the conditional prompt rule retire entirely) — recorded in the RFC as future work.

### Behaviors a reader should expect (beyond the headline)
- Over-cap or transport-truncated diffs degrade to a summary chip (filename, −N/+M, truncation note) that expands the details panel — a new affordance, not the old collapsed rendering.
- Each card folds to its chip and back via the header chevron; fold state persists across virtualized scrolling within the session.
- A dashboard-started session continued from a channel (Slack/Discord/...) gets a per-turn re-assertion of the hard ```diff mandate on non-dashboard turns; it is never retracted in the dashboard direction.

### CODE REVIEW

**Comment by CrysisDeu** [comment]
## Disposition — UX Review (lane: ux-review)

**Finding 1 (over-cap edits lose their only visible diff): ADOPTED, fixed in 78eba5d8811f20b6b689f85f3142b9af81c65744.** `presentToolDiff` now degrades an over-cap edit to a summary chip — filename, −N/+M counts, click expands the details panel holding the full diff — instead of returning null. Every edit-tool diff has a visible transcript trace regardless of size. Covered by `toolDiff.test.tsx` ("degrades an over-cap diff to a summary — never to nothing") and a ToolCallLine render test asserting the chip appears and the full card does not.

**Finding 2 (cards exempt from collapseAll with no override): ADOPTED, fixed in 78eba5d8811f20b6b689f85f3142b9af81c65744.** `isDiffCardItem` was removed from `isVisibleInline`: collapseAll (the user's explicit fold preference) now folds diff rows with the other working steps; only DEFAULT mode keeps them out of the tool-call fold. `TurnBlock.test.tsx` re-pinned to the new contract (step counter includes the edit row under collapseAll).

**Comment by CrysisDeu** [comment]
## Disposition — Design Review (lane: design-review)

**Finding 1 (channel regression rides on per-turn model judgment): ADOPTED, fixed in 80383257a72ccbeaf8ccf3879d70dbfa8bfb8c30.** Implemented your suggestion: the runtime clause is now decided SERVER-SIDE. `_critical_rules_for` (context.py) reuses the trusted `_runtime_display_name` resolution to select between two fixed module constants — dashboard sessions get the don't-repeat rule, every other surface (Slack/Discord/Telegram/cron/subagent/CLI) gets the original unconditional mandate verbatim. Only the tool-vs-shell clause (a) stays with the model, and its failure mode is a duplicate diff, never a missing one. Covered by `test_diff_rule_is_runtime_selected` (explicit `runtime_source` wins over the key-derived guess).

**Finding 2 (over-cap edits lose their primary display): ADOPTED, fixed in 78eba5d88 (also raised by UX Review).** Over-cap edits degrade to a summary chip (filename, −N/+M, expands the details panel) instead of nothing.

**Finding 3 (temp-screenshots committed at repo top level): NOT ADOPTED.** Committed QA media under `temp-screenshots/<feature>/` is this repo's documented convention (kirocrew-worktree-dev Rule 9 names it as the one sanctioned committed deliverable; ~2,000 files already live there, as First Principles' review of this same PR counted). The PR-body links are commit-SHA-pinned and require the files on the merged history. Deviating from the convention in one PR would break the screenshot-evidence pattern the other lanes verify.

**Comment by CrysisDeu** [comment]
## Disposition — First Principles Review (lane: first-principles-review)

**Watch (RFC ships self-contradicting): ADOPTED, fixed in 80383257a72ccbeaf8ccf3879d70dbfa8bfb8c30.** Front matter now `status: in-progress`, `implementation-prs: ["#5012"]`; the index row states what ships with this PR and names the unstarted §3.3 remainder.

**Watch (RFC §3.1/§5 contradict the shipped code): ADOPTED, fixed in 80383257a72ccbeaf8ccf3879d70dbfa8bfb8c30.** §3.1 rewritten to the actual architecture (presentation renders inside ToolCallLine — one site covers ChatPage, ChatPane, and app-sdk surfaces — plus the summary-chip degradation); the §5 `complete={false}` shimmer row replaced with the true statement (live toolLog input and persisted meta.input carry the same redacted value, so no truncation asymmetry exists).

**Subtraction (un-export `DIFF_CARD_MAX_LINES`): ADOPTED, fixed in 80383257a72ccbeaf8ccf3879d70dbfa8bfb8c30.** Module-private; tests inline the boundary values.

**Comment by CrysisDeu** [comment]
## Disposition — Design Review round 2 (lane: design-review)

**Finding 1 (edits outside the built-in tools lose their trace under the relaxed rule): ADOPTED, fixed in 76ac00ff0b4f805c4117c999be05cf0db327d696.** The dashboard rule now scopes the don't-repeat clause to "the BUILT-IN file-editing tools" and explicitly lists "an MCP tool that writes files" in the must-emit set (`_DIFF_RULE_DASHBOARD`, `prompt.md`, `prompt-orchestrator.md`). Which tool it invoked is something the model directly observes, so this keeps the judgment on an observable — an MCP or adapted-harness edit gets the model-authored diff block exactly as before this PR.

**Finding 2 (6 KB `make_unified_diff` cap vs the chip's full-diff promise): ADOPTED, fixed in 76ac00ff0b4f805c4117c999be05cf0db327d696.** The cap defaults to 64 KiB — sized so the full-card range (≤400 lines) is never cut — and truncation now lands on a LINE boundary with a `\ diff truncated` annotation (the unified-diff escape convention; renderers skip it), so a cut is machine-detectable instead of a silent mid-line slice. `client.py`'s duplicate copy now delegates to the single implementation. Covered by `test_truncation_cuts_at_line_boundary_and_is_marked` / `test_under_cap_diff_is_not_marked`; the residual (counts understate past 64 KiB, marked) is documented in the RFC risk table.

**Suggestion (compare raw source, not localized display): ADOPTED, fixed in 76ac00ff0b4f805c4117c999be05cf0db327d696.** Resolution extracted to `_resolve_runtime_source` (raw key); `_runtime_display_name` is now a display mapping over it and `_critical_rules_for` compares the raw source — no dependency on `_RUNTIME_DISPLAY` string stability.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 2 (lane: ux-review)

**Finding 1 (denied edits look applied): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72.** A rejected or auto-denied call never promotes: `diffView` is gated on `!isRejected && !isAutoDenied`. The full diff stays readable in the expanded details panel; the pill's red/amber status icon is no longer visually contradicted by a first-class card. Covered by "suppresses the card for a rejected edit".

**Finding 2 (collapseAll loses the visible change record): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72, exactly as you suggested.** Under "collapse all working steps" every diff row demotes to the one-line −N/+M summary chip (new `useCollapseAllSteps` hook; `presentToolDiff` takes a card-line ceiling and `0` forces summary), and `TurnBlock` keeps diff rows visible-inline again — one visible line per change, no wall of cards, chip click expands the full diff. Covered by "demotes the card to the summary chip under collapse-all-steps" and the TurnBlock visibility test.

**Comment by CrysisDeu** [comment]
## Disposition — First Principles Review round 2 (lane: first-principles-review)

**Subtraction (inline +/− loop duplicates `countDiffStats`): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72.** `presentToolDiff` imports `countDiffStats` from `utils/diffLineCounts` — thanks for the pointer, the module was unknown to this change.

**Subtraction (`headerPath` duplicates `extractFilePath`): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72.** `extractFilePath` is exported from `DiffBlock.tsx` and reused; the local scanner is deleted.

**Subtraction (one-member `_DIFF_CARD_RUNTIMES` frozenset): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72.** Direct `source == "dashboard"` comparison.

**Watch (prompt.md still ships the model-judged runtime clause, three spellings to sync): ADOPTED, fixed in e5c5f31ee475886727431a0e585564835a575b72.** Clause (b) is deleted from both bundled prompts; the paragraph now defers to the runtime-selected critical rules as authoritative and keeps only the tool-vs-shell/MCP distinction (the one thing the model can observe). One authoritative spelling per surface, selected server-side.

**Watch (AGENTS.md: new markdown file needs explicit instruction): CONFIRMED-AS-INSTRUCTED.** The RFC was explicitly requested by the repository owner in the originating session ("写一个 RFC 把方案定下来" — "write an RFC to settle the design"), not convention-inherited.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 3 (lane: ux-review)

**Finding 1 (truncated diffs look complete): ADOPTED, fixed in 820bf329507722427d06b65973496599e95d8e76.** `presentToolDiff` detects the transport's `\ diff truncated` annotation: a truncated payload never renders as a card (it can sit under the line cap while missing most of the change) — always the summary chip, now carrying a visible localized "diff truncated" note (new `pages.chat.toolCallLine.diff_truncated` key, all 14 catalogs) so the −N/+M counts read as lower bounds. Covered by "a transport-truncated diff is always a summary, flagged truncated".

**Finding 2 (edit-kind row without a unified diff leaves no trace): ADOPTED, fixed in 820bf329507722427d06b65973496599e95d8e76.** A `kind === 'edit'` row whose input is not diff-shaped falls back to a pathname chip (basename from the existing arg-path extraction, expands the details panel); counts are omitted because nothing exists to count them from. Covered by "falls back to a pathname chip for an edit row without a transported diff".

**Comment by CrysisDeu** [comment]
## Disposition — Design Review round 3 (lane: design-review)

**Finding (cross-surface sessions outrun the session-start rule selection): ADOPTED, fixed in 820bf329507722427d06b65973496599e95d8e76, using your asymmetric suggestion.** The per-turn runtime refresh in `build_message` (the same path that re-injects `[RUNTIME]` from trusted dispatcher metadata) now appends the hard channel mandate whenever the resolved source for THIS turn is not `dashboard` — and never retracts the mandate in the other direction, exactly because the fail direction is asymmetric (dashboard duplicate = cosmetic; missing channel diff = no record). `prompt.md` / `prompt-orchestrator.md` no longer call the session-start rules "authoritative"; they now defer to "the critical rules injected for your session and any per-turn surface note next to the [RUNTIME] line — the most recent instruction wins", so the model is never told to disbelieve the refreshed runtime. Covered by `test_channel_turn_reasserts_diff_mandate_mid_session` (discord turn injects the mandate; dashboard turn injects nothing).

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 4 (lane: ux-review)

**Finding 1 (no per-card collapse): ADOPTED, fixed in 3bef27a9f7a024b83c5a6e0e02a81f3468777f45.** Every diff presentation now has a one-line chip handle: for a promoted card the chip folds the card away after reading and unfolds it again (local state, default open — the promotion's whole point); for summary/pathname rows it expands the details panel as before. Covered by "the chip folds an open card and unfolds it again".

**Finding 2 (truncated counts read as exact): ADOPTED, fixed in 3bef27a9f7a024b83c5a6e0e02a81f3468777f45.** Truncated transports render counts as `≥-N ≥+M` next to the localized "diff truncated" note — the numbers now read as the lower bounds they are.

**Suggestion (identical chips for same-named files): ADOPTED, fixed in 3bef27a9f7a024b83c5a6e0e02a81f3468777f45.** The chip carries the full path in its native tooltip, matching the DiffBlock wrapper's existing convention.

Also fixed this round: the CI "Frontend Lint & Type Check" red was the pseudolocale gate — `en-XA.json` is generated and my hand-inserted key didn't match `gen-pseudolocale.mjs` output; regenerated and committed, `i18n:check` fully green locally.

**Comment by CrysisDeu** [comment]
## Disposition — First Principles Review round 3 (lane: first-principles-review)

**Watch (default experience regresses to a chip under `collapseAllSteps: true`): ADOPTED — root-caused and fixed in 5d4db7db3dfdc6f5823db9b92e1244f560900fef.** You were right that the demotion made the PR's headline claim false for default-config users. The fix removes the collapseAll→chip demotion entirely: the card renders regardless of that preference, because a file change is a RESULT, not a working step — the same class as the prose diff the final summary used to carry, which the preference never folded either. The `useCollapseAllSteps` hook is deleted; `presentToolDiff` loses its `maxCardLines` override. Default users now see the full card, matching the screenshots.

**Subtraction (drop the per-card fold chip — `collapseAllSteps` already provides the relief): ADOPTED IN INVERSE, fixed in 5d4db7db3dfdc6f5823db9b92e1244f560900fef.** Your redundancy argument is correct — two density mechanisms is one too many — but the one to delete is the collapseAll demotion (it caused your own Watch finding above), not the per-card fold: per-card fold gives relief without regressing the default, and UX explicitly requested it. One mechanism remains.

**Subtraction (drop the `editWithoutDiff` pathname chip): NOT ADOPTED.** The risk-table sentence you cite ("nothing is lost") describes the `tool_kind` MISLABEL case, where the relaxed rule's premise fails closed. The pathname chip covers a different case: kind IS `edit` (the relaxed rule applies, the model stays silent) but no diff content block arrived — there the zero option is an invisible file change, which UX round 3 flagged as breaking the PR's core promise. The comparator is post-relaxation behavior, not today's.

**Watch (RFC §3.1 contradicts shipped collapseAll behavior): ADOPTED, fixed in 5d4db7db3dfdc6f5823db9b92e1244f560900fef** — §3.1 rewritten to the both-folds-excluded + per-card-fold design, and the companion behaviors (cap raise, fold chip, pathname chip, rejected suppression, per-turn channel mandate) are now DECLARED in §3.1 rather than riding along.

**Watch (RFC presents the 64 KiB cap as pre-existing): ADOPTED, fixed in 5d4db7db3dfdc6f5823db9b92e1244f560900fef** — the risk table now states "this PR raises `make_unified_diff`'s cap from 6 000 chars to 64 KiB".

Also this round: the CI Frontend Lint red was the i18n [added-lines] gate flagging the frontend's copy of the truncation sentinel string; it is a wire pattern, not user copy, and is now a regex literal (the rule checks string literals), which also states the protocol role precisely.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 5 (lane: ux-review)

**Finding 1 (chip/card double header): ADOPTED, fixed in ec6384173ba7a76962a391e7287a6192960ca92d.** The fold control now lives in the card's OWN header row (`DiffBlock` gains an optional `onFold`, rendered as a chevron next to the split-view/copy controls); the chip renders only when the card is folded, as its re-open handle. No fact renders twice. Covered by the updated fold test ("An OPEN card shows no chip").

**Suggestion (chip aria-label collides with the pill's): ADOPTED, fixed in ec6384173ba7a76962a391e7287a6192960ca92d.** Card-mode fold/unfold uses a distinct localized `aria_show_diff` / `aria_hide_diff` pair (all 14 catalogs); the details-panel chips keep the existing pair.

**Finding 2 (collapse-all users still get open cards): NOT ADOPTED — position stated, escalated to the repository owner.** This is the third round of reversals on the same design point across two advisory lanes: UX r1 asked collapseAll to fold cards; the demotion built for that caused First Principles r3's finding that DEFAULT-config users (collapseAllSteps ships `true`) regressed from today's full prose diff to a chip — the PR's headline claim held only for a non-default setting. The current design resolves that conflict deliberately: a file change is a RESULT, not a working step — the same class as the prose ```diff the final summary carries today, which the collapseAll preference has never folded either. So the preference's contract ("fold my working steps") is not violated by an open diff card any more than it is violated today by a prose diff in the assistant message. Per-card fold (your r4 request, now in the card header) is the density mechanism, applied by the reader where it is wanted rather than globally. The test-comment contradiction you cite was real and is fixed in ec6384173ba7a76962a391e7287a6192960ca92d. Whether collapseAll should ALSO start diff cards folded is a genuine product preference question with no objectively right answer left — it is escalated to the repository owner on this PR rather than flip-flopped a fourth time between lanes.

**Comment by CrysisDeu** [comment]
## Response — GPT 5.6 Review blocking findings (both fixed in e0c0d831fe69be07137049c200f461ae6ff4569c)

**BLOCKING 1 (diff-less edit events lose the change display): FIXED.** Rather than retaining the unconditional mandate (which would reinstate the token/duplication cost the PR removes), the frontend now DERIVES a unified diff from the bare-JSON edit arguments themselves — a create's `fileText`/`content` renders as an all-additions card (`/dev/null` old side), an `oldStr`/`newStr` pair renders as a replace hunk (`deriveJsonEditDiff` in `toolDiff.ts`). The changed content displays as a first-class diff for every recognizable edit shape; only a payload carrying no content at all falls back to the pathname chip. Covered by "derives a diff from bare-JSON edit arguments".

**BLOCKING 2 (three sibling buttons in the diff header): FIXED.** The fold button is removed from the actions row. The fold affordance is now the header strip itself — a transparent full-width button UNDER the actions group (`z-0` vs the controls' `z-10`), which is not a sibling in the horizontal button group and matches the UX lane's round-5 suggestion ("make the card's header row the fold handle"). The actions row is back to Open + copy (+ the view-mode toggle, which the rule's own exemption list names).

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 6 (lane: ux-review)

**Finding 1 (hidden fold affordance): ADOPTED, fixed in 9500771653b65b4f9953379e85b623079475d4c6.** The header fold handle now reveals a chevron on card hover (the same reveal convention the actions row uses) instead of being permanently invisible.

**Finding 2 (invisible focus stop): ADOPTED, fixed in 9500771653b65b4f9953379e85b623079475d4c6.** The control carries `focus-visible:ring-2 focus-visible:ring-accent/50` and the chevron becomes visible on keyboard focus, matching the sibling chip's focus treatment.

**Comment by CrysisDeu** [comment]
## Disposition — First Principles Review round 4 (lane: first-principles-review)

**Subtraction (drop the `editWithoutDiff` pathname chip — `FileChangeChips` already carries the trace): ADOPTED, fixed in 9500771653b65b4f9953379e85b623079475d4c6.** Verified your grep: `FileChangeChips` renders from the `file_changes` snapshot whose producer (`_snapshot_write_target`) covers exactly the `create` / `strReplace` / `insert` shapes, and it is fold-proof. The chip, its branch, and its test are deleted; the RFC's companion-behaviors list now names `FileChangeChips` as the trace for unrecognizable shapes. (Note the case has also narrowed since your review base: `derive_edit_diff` now converts those same shapes server-side, so a non-diff edit input is rarer still.)

**Watch (cap raise multiplies persisted meta ~11×, undeclared in the description): ADOPTED.** The PR description's Changes section now declares the 6 000 → 64 KiB raise with its payload consequence and the 1 MB ceiling it sits under.

**Subtraction (drop the per-card fold, let collapseAll fold instead): NOT ADOPTED — deferred to the owner, and your own framing agrees it is the weaker call.** This is the collapseAll design point already escalated to the repository owner on this PR (UX r5 disposition); the parity argument you acknowledge as real ("the prose diff never folded either") is the standing position until the owner rules. Reversing it lane-to-lane a fourth time is the one move everyone loses.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 7 (lane: ux-review)

**Finding (fold overlay's `title` swallows the full-path tooltip): ADOPTED, fixed in 6698bb20905641b099c01d2a76bc8a6d991afc7b.** The overlay's `title` is removed — the wrapper's full-path tooltip now shows across the whole header again (the one signal telling two same-named files apart), and `aria-label` continues to carry the action for assistive tech. The hover chevron from the previous round remains the sighted-discovery affordance.

**Suggestion (screenshots lack the shell negative case): PARTIALLY ADOPTED.** The PR-body claim is reworded to match exactly what the screenshots show; the negative case is covered by automated tests (`toolDiff.test.tsx` "does not render a card for a shell tool with diff-shaped input", TurnBlock's execute-kind test) rather than by re-shooting the isolated-gateway evidence for each round.

**Comment by CrysisDeu** [comment]
## Disposition — First Principles Review round 5 (lane: first-principles-review)

**Watch (content-only derivation renders an existing-file edit as pure additions — a misleading record): ADOPTED, narrowed in 6698bb20905641b099c01d2a76bc8a6d991afc7b.** `derive_edit_diff` now derives from content keys ONLY for `command == "create"`, where the additions ARE the whole change, exactly. `insert` and unknown shapes derive nothing (new regression test pins this) and keep their fold-proof trace via the `file_changes` snapshot channel. The RFC's companion-behaviors list now declares the derivation and its create-only scope.

**Subtraction (shrink to strReplace parity; let the `editWithoutDiff` chip be the one fallback): NOT ADOPTED AS WRITTEN — your review base is stale and the proposal contradicts your own round 4.** This review reads `87c388d66`; the pathname chip you designate as "the one fallback" was DELETED in `950077165` — at your own round-4 direction ("Drop the `editWithoutDiff` pathname chip — `FileChangeChips` already carries that trace"), which I adopted. The create-only narrowing above takes your round-5 point where it is right (misleading additions rendering) without resurrecting the mechanism you had me remove.

**Watch + Subtraction (item 4 / fold machinery exists only because of the collapseAll exemption): NOT ADOPTED — this is the escalated design point.** Same disposition as rounds 3-4: the collapseAll question is with the repository owner; until ruled, the parity position stands and the fold machinery is UX's requested (and since round 6, accessibility-hardened) density control. One authoritative decision beats a fifth lane-to-lane reversal.

**Comment by CrysisDeu** [comment]
## Disposition — Design Review round 4 (lane: design-review)

**Finding 1 (insert edits lose their diff; the create-only rationale is wrong for insert): ADOPTED, fixed in 55d97a6ba99a195c9bb798d73206563234b393ed.** You are right — an insert IS additions-only, so the "misrepresents existing content" rationale did not apply to it. `derive_edit_diff` now derives a positioned addition hunk for a line-numbered `insert` (`@@ -N,0 +N+1,M @@`); an insert WITHOUT a line number still derives nothing (the hunk position would be a guess) and keeps its trace via the `file_changes` snapshot. Two regression tests pin both branches; the RFC and comments state the exact scope.

**Finding 2 (committed screenshot binaries): NOT ADOPTED — same disposition as your round 3.** Committed QA media under `temp-screenshots/<feature>/` is this repository's documented convention (worktree guide Rule 9 names it the sanctioned committed deliverable; ~2 000 files precede this PR). Commit-SHA-pinned same-origin links are also the repo's documented screenshot mechanism — PR-uploaded images live on a different origin that the repo's own guidance forbids. A convention change is out of this PR's scope.

**Suggestion (symmetric per-turn dashboard note for channel-started sessions): DEFERRED, as your own framing suggests** ("only if the duplicate proves annoying in practice"). The asymmetry is deliberate: the failure directions are not symmetric, and a duplicate diff is cosmetic.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 8 (lane: ux-review)

**Finding 1 (focus lost on every fold/unfold): ADOPTED, fixed in 55d97a6ba99a195c9bb798d73206563234b393ed.** The toggle hands focus to its counterpart: both controls carry `data-diff-toggle`, and an effect keyed on the fold flip focuses whichever half mounted (chip after fold, header control after unfold). Keyboard users keep their place.

**Finding 2 (folds don't survive scrolling): ADOPTED, fixed in 55d97a6ba99a195c9bb798d73206563234b393ed.** Fold state is remembered at module scope by `tool_call_id` (`foldedDiffCards`, the same page-lifetime pattern the file's `revealedToolIds` already uses), so virtualizer recycling no longer reopens folded cards. Pinned by "a fold survives unmount/remount (virtualized transcript)".

**Suggestion (turn-level "fold all diffs"): DEFERRED** — a bulk affordance is a reasonable follow-up once the per-card mechanics settle; noted for the RFC's future-work section rather than grown mid-review.

**Comment by CrysisDeu** [comment]
## Disposition — UX Review round 9 (lane: ux-review)

**Finding 1 (whole-header invisible click-trap): ADOPTED, fixed in bd225f1f67d0680c46afad1b00e7b635b31555fe.** The fold target shrinks to a 32px chevron zone at the header's left edge — the filename is inert again (selectable, copyable, tooltip intact).

**Finding 2 (chip `aria-label` erases visible content): ADOPTED, fixed in bd225f1f67d0680c46afad1b00e7b635b31555fe.** The chip's `aria-label` is removed entirely: the accessible name is now its visible content (filename + −N/+M), which restores voice-control matching ("click app.py"), and `aria-expanded` carries the toggle state. This also supersedes the round-5 label pairing on the chip — name-from-content is distinct by construction; the icon-only header control keeps its explicit label.

**Suggestion (chevron at rest): ADOPTED, fixed in bd225f1f67d0680c46afad1b00e7b635b31555fe.** The chevron renders at rest in muted color (brightening on hover/focus) so the only density control is discoverable without mouseover.

**Comment by CrysisDeu** [comment]
## Disposition — Design Review round 5 (lane: design-review)

**Finding 1 (five prose copies of the diff rule can drift): ADOPTED, reduced in 28c6b0d3f87f404ca3a630c0188fbadea0639280.** Both bundled prompts no longer restate the per-surface rule at all — they now defer to the injected critical rules + per-turn note and keep only the FORMAT specification (headers, `/dev/null`), which is surface-invariant. The normative copies drop from five to three, all in `context.py` (`_DIFF_RULE_DASHBOARD`, `_DIFF_RULE_CHANNEL`, the per-turn re-assertion), i.e. all server-side and colocated in one module — a prompt edit can no longer drift against a sibling it cannot see.

**Finding 2 (dashboard fidelity for non-tool edits rides on model self-classification): ACKNOWLEDGED AS RESIDUAL, no change.** This is inherent to the design: the runtime cannot observe HOW a file was changed, so the tool-vs-shell distinction is the one clause that must stay with the model — the RFC's risk table names exactly this with its fail direction (worst case: a `file_changes` chip instead of a prose diff, never zero trace). "Watch in edit-heavy real sessions" is the right posture; if real-session data shows misclassification, the follow-up is the `OutputEvent` extension that retires prompt reliance entirely (RFC §3.3).

**Comment by CrysisDeu** [comment]
## Response — GPT 5.6 blocking finding (fixed in 6950ee565e8216143db8dff8f8c40cbac0b80bac)

**Malformed edit args abort the turn: FIXED.** `derive_edit_diff` type-narrows every argument it forwards: a non-string `path` returns `""` outright, and non-string `oldStr`/`newStr` coerce to `""` before `difflib` — a TypeError can no longer escape the dispatch path. Pinned by `test_non_string_arguments_never_reach_difflib` (numeric path, dict oldStr, list path, numeric fileText). The create/insert branches already isinstance-gated their content values.

**Comment by CrysisDeu** [comment]
**Disposition — First Principles CONCERNS (on 6950ee565):**

1. **Subtraction: delete `aria_show_diff` (0 consumers)** — ADOPTED. Removed
   from `en.manual.json` and all 12 hand-maintained locales (pure line-delete),
   `en-XA.json` regenerated. The dead-keys ratchet independently caught the
   same key and is back at baseline.
2. **Items 2/3/6 undeclared in the PR body** — ADOPTED. The body now carries a
   "Behaviors a reader should expect" section declaring the summary-chip
   affordance (correcting the stale "keep the collapsed rendering" phrasing),
   per-card fold with scroll-persistent state, and the per-turn channel
   re-assertion of the hard diff mandate.

**Comment by CrysisDeu** [comment]
**Fix note for the blocking finding (head 72848194a48b8ba26e91d1e5f65233ceb270e1a4):**

- **BLOCKING — header paths truncated at spaces** — fixed at the root instead
  of removing the affordance: `extractFilePath`'s terminator is now TAB-only
  (the unified-diff timestamp separator; both difflib and git emit the path as
  the whole remainder of the line, so a space is a legal path character).
  `/work/report final.md` now resolves exactly; a space-separated timestamp
  from a foreign tool stays attached, fails the existence probe, and the
  open affordance is simply not offered — fail-safe in the harmless direction.
  Regression tests cover the spaced-sibling case and TAB-timestamp stripping.
  (The regexes predate this PR, but the promotion raised their blast radius —
  the finding stands and the shared fix benefits every consumer.)
- **FINDING — local imports in client.py** — fixed: `make_unified_diff` and
  `derive_edit_diff` moved into the existing top-level `_dispatch` import
  block; no cycle (`_dispatch` does not import `client`).

**Comment by CrysisDeu** [comment]
**Fix note for the blocking finding (head 79fe604e8):**

- **BLOCKING — minimal-context crons lose the file-diff mandate** — fixed
  exactly as suggested: both bundled prompts now default to the unconditional
  ```diff mandate, with the injected critical rules (or a per-turn [RUNTIME]
  surface note) acting as the RELAXATION rather than the source of the
  requirement. A session that receives no critical-rules block at all — the
  `minimal_context=True` cron early-return — falls back to the hard mandate
  instead of silence. The pointer structure survives (the prompts still do not
  restate the per-surface rule); only the fail-safe direction flipped. RFC
  §3.2 records the default direction and the minimal-context rationale.

**Review by bolichen97 — APPROVED** [review]
